src/replication.c

```diff
@@ -750,6 +766,168 @@ is_replica_newly_connected(spec_t *spec, replica_t *new_replica)
+		rc = -1;
+		goto error_in_update;
+	}
+	//TODO: Improve read code
+	read_data = (char *)malloc(sizeof(char) * 128);
+	tmp_data = (char *)malloc(sizeof(char) * 64);
+	rc = istgt_readline_socket(fd, read_data, 128, tmp_data, 64, &tmpidx, &tmpcnt, rtimeout);
+	if (rc < 0) {
+		ISTGT_ERRLOG("failed to read the data on %d", fd);
+		goto clean_data;
+	}
+	if(strcmp(read_data, err) == 0) {
+		ISTGT_ERRLOG("failed to update %s", data);
+		rc = -1;
+		goto clean_data;
+	}
+	rc = 1;
+clean_data:
+	free(read_data);
+	free(tmp_data);
+error_in_update:
+	close(fd);
+	return rc;
+}
+
+static int
+update_known_replica_list(spec_t *spec, replica_t *rep) {
+	int known_replica_count, data_len, rc;
+	char *data;
+	const char *json_string;
+	struct json_object *jobj;
+
+	known_replica_count = get_known_replica_count(spec);
+
+	if (known_replica_count >= spec->desired_replication_factor) {
+		ISTGT_ERRLOG("current replica count %d is greater "
+		    "than or equal to desired replciation factor %d",
+		    known_replica_count, spec->desired_replication_factor);
+		return -1;
+	}
+
+	jobj = json_object_new_object();
+	json_object_object_add(jobj, "replicaId",
+	    json_object_new_uint64(rep->zvol_guid));
+	json_string = json_object_to_json_string_ext(jobj, JSON_C_TO_STRING_PLAIN);
+
+	data_len = strlen(json_string) + 1;
+	data = (char *) malloc(sizeof(char *)*data_len);
+	memset(data, 0, data_len);
+	strncpy(data, json_string, data_len);
+
+	rc = update_persistent_volume(data);
+	if (rc < 0) {
+		ISTGT_ERRLOG("failed to update replica{%lu} to known replica list", rep->zvol_guid);
+	} else {
+		spec->known_replicas_uid[++known_replica_count] = rep->zvol_guid;
+		rep->updated_as_known_replica = true;
+	}
+
+	free(data);
+	json_object_put(jobj);
+	return rc;
+}
+
+#define check_is_it_known_replica(_spec, _rep, val)		\
+	do {							\
+		int i;						\
+		val = false;					\
+		for(i=0; i<_spec->replication_factor; i++) {	\
+			if (_spec->known_replicas_uid[i] ==	\
+			    _rep->zvol_guid)			\
+				val = true;			\
+			if (_spec->known_replicas_uid[i] == 0)	\
+				break;				\
+		}						\
+	}while(0)
+/*
+ * is_known_replica returns true whether the replica is known or not based
+ * on below factors
+ * 1) If quorum is off, then treat it as a known replica.
+ * 2) If quorum is on, and replicas checkpointed nums is zero then
+ *    treat it as a known replica.
+ * 3) If quorum is on and its checkpointed nums are not zero check whether
+ *    current replica is present in known list, if it present treat as known
+ *    replica else unknown replica.
+ */
+static bool
+is_known_replica(spec_t *spec, replica_t *new_replica)
+{
+	int known_replica_count;
+	bool result = true;
+
+	if (new_replica->quorum == 0) {
+		return result;
+	}
+	if (new_replica->initial_checkpointed_io_seq == 0) {
+		return result;
+	}
+	//TODO: This is to handle upgrade scenarios should we handle
+	//  in code or in upgrade process?
```

Comment:
we will handle during upgrade..

Reply:
so no need of this

Reply:
Code has taken care but naming conventions need to update